### PR TITLE
Use `BTreeSet`s instead of `Vec`s for unique, ordered types

### LINKED ISSUE #187: Use BTreeSets when ordered sets are required
## Description

Some places, like Features/Unlock Conditions in outputs, use `Vec` to maintain proper ordering. However, these collections are unique by type, and the collection type should represent that. `BTreeMap` is suitable for this purpose, and the order can be specified idomatically using `Ord`.

## Motivation

Code simplification and bugs are less likely when using the guarantees provided by types that are purpose-made.

## Open questions (optional)

- Is there any reason `Ord` could have different impls for the contained types based on context?

## Are you planning to do it yourself in a pull request?

Yes

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Uses `BTreeSet`s when possible to avoid some work when building types.

## Links to any relevant issues

- Closes #187 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)

### CODE REVIEW

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/types/block/payload/milestone/option/mod.rs
Comment (thibault-martinez) [inline sdk/src/types/block/payload/milestone/option/mod.rs:1]: Changelog entries

File: sdk/src/types/block/parent.rs
Comment (thibault-martinez) [inline sdk/src/types/block/parent.rs:1]: Changelog entries

File: sdk/src/types/block/output/unlock_condition/mod.rs
Comment (thibault-martinez) [inline sdk/src/types/block/output/unlock_condition/mod.rs:1]: Changelog entries

File: sdk/src/types/block/output/native_token.rs
Comment (thibault-martinez) [inline sdk/src/types/block/output/native_token.rs:1]: Changelog entries

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/types/block/payload/milestone/option/mod.rs [hunk 1]
```diff
@@ -92,6 +101,17 @@ impl MilestoneOptions {
         Ok(Self(milestone_options))
     }
 
+    /// Creates a new [`MilestoneOptions`] from an ordered set.
+    pub fn from_set(milestone_options: BTreeSet<MilestoneOption>) -> Result<Self, Error> {
```
Comment (thibault-martinez) [inline sdk/src/types/block/payload/milestone/option/mod.rs:116]: Does that not need a specific Ord impl or something? How are they ordered?

File: sdk/src/types/block/parent.rs [hunk 2]
```diff
@@ -31,8 +31,8 @@ impl Parents {
     /// The range representing the valid number of parents.
     pub const COUNT_RANGE: RangeInclusive<u8> = 1..=8;
 
-    /// Creates new [`Parents`].
-    pub fn new(mut inner: Vec<BlockId>) -> Result<Self, Error> {
+    /// Creates new [`Parents`] from a vec.
+    pub fn from_vec(mut inner: Vec<BlockId>) -> Result<Self, Error> {
         inner.sort_unstable_by_key(|a| a.pack_to_vec());
```
Comment (thibault-martinez) [inline sdk/src/types/block/parent.rs]: Not related to your PR but I'm trying to remember why this is a `pack_to_vec`. I feel like I remember `sort_unstable_by_key` couldn't work with references or something.It feels a bit useless. Maybe see if you can avoid it?

File: sdk/src/types/block/output/nft.rs [hunk 3]
```diff
@@ -94,10 +94,10 @@ impl NftOutputBuilder {
         self
     }
 
-    /// Adds an [`UnlockCondition`] to the builder.
+    /// Adds an [`UnlockCondition`] to the builder, if one does not already exist of that type.
```
Comment (thibault-martinez) [inline sdk/src/types/block/output/nft.rs:97]: I'm not sure I understand this comment. Maybe say it replaced it it was already there?

File: sdk/src/types/block/payload/milestone/option/mod.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/payload/milestone/option/mod.rs:116]: Surprised the compiler didn't catch this, thanks.

Comment (thibault-martinez) [reply sdk/src/types/block/payload/milestone/option/mod.rs:116]: I'm still not sure how this is compiling tbh

Comment (thibault-martinez) [reply sdk/src/types/block/payload/milestone/option/mod.rs:116]: Does `BTreeSet` not have an Ord bound?

File: sdk/src/types/block/output/nft.rs (hunk 3 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/output/nft.rs:97]: It doesn't replace it, that's what `replace_unlock_condition` does.

Comment (thibault-martinez) [reply sdk/src/types/block/output/nft.rs:97]: Oh my bad, I think I misread the doc

File: sdk/src/types/block/payload/milestone/option/mod.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/payload/milestone/option/mod.rs:116]: No, only the fns do.

File: sdk/CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +21,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## 0.3.0 - 2023-XX-XX
 
+### Added
+- `NativeTokensBuilder::finish_set`;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md:26]:
```suggestion

- `NativeTokensBuilder::finish_set`;
```

File: sdk/CHANGELOG.md [hunk 5]
```diff
@@ -29,6 +33,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `PostBlockPayload::payload_dto` renamed to `payload`;
 - `SendNativeTokens::addresses_native_tokens` renamed to `addresses_and_native_tokens`;
 - `SendNft::addresses_nft_ids` renamed to `addresses_and_nft_ids`;
+- `Output` builder types, `NativeTokensBuilder`, and `Burn` now use unique, ordered sets for unlock conditions, features, and native tokens. `add_x` and `replace_x` methods thus function appropriately;
+- `Features`, `UnlockConditions`, `NativeTokens`, `MilestoneOptions`, and `Parents` constructor `new` renamed to `from_vec`;
+- Modified `Ord` and `PartialOrd` implementations for `Feature`, `UnlockCondition`, and `NativeToken` to support unique, ordered sets;
```
Comment (thibault-martinez) [inline sdk/CHANGELOG.md]: Also `MilestoneOption`

**Review by thibault-martinez — DISMISSED** [review]

File: sdk/src/types/block/output/foundry.rs [hunk 6]
```diff
@@ -272,12 +249,12 @@ impl From<&FoundryOutput> for FoundryOutputBuilder {
...
         Self {
             amount: OutputBuilderAmount::Amount(output.amount),
-            native_tokens: output.native_tokens.to_vec(),
+            native_tokens: output.native_tokens.iter().copied().collect(),
             serial_number: output.serial_number,
             token_scheme: output.token_scheme.clone(),
-            unlock_conditions: output.unlock_conditions.to_vec(),
-            features: output.features.to_vec(),
-            immutable_features: output.immutable_features.to_vec(),
+            unlock_conditions: output.unlock_conditions.iter().cloned().collect(),
```
Comment (kwek20) [inline sdk/src/types/block/output/foundry.rs:255]: cant we clone these normally?

File: sdk/src/types/block/output/basic.rs [hunk 7]
```diff
@@ -197,9 +183,9 @@ impl From<&BasicOutput> for BasicOutputBuilder {
     fn from(output: &BasicOutput) -> Self {
         Self {
             amount: OutputBuilderAmount::Amount(output.amount),
-            native_tokens: output.native_tokens.to_vec(),
-            unlock_conditions: output.unlock_conditions.to_vec(),
-            features: output.features.to_vec(),
```
Comment (kwek20) [inline sdk/src/types/block/output/basic.rs:202]: and clone here

File: sdk/src/types/block/output/foundry.rs (hunk 6 above)
Comment (thibault-martinez) [reply sdk/src/types/block/output/foundry.rs:255]: It's not the same type

File: sdk/src/types/block/output/feature/mod.rs [hunk 8]
```diff
@@ -188,6 +208,17 @@ impl Features {
         Ok(Self(features))
     }
 
+    /// Creates a new [`Features`] from an ordered set.
+    pub fn from_set(features: BTreeSet<Feature>) -> Result<Self, Error> {
```
Comment (kwek20) [inline sdk/src/types/block/output/feature/mod.rs:212]: better to impl it as IntoIter<> i thnk

Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/mod.rs:212]: But then you lose the ordered/unique properties

Comment (kwek20) [reply sdk/src/types/block/output/feature/mod.rs:212]: then we wouldnt need this method if we just implement it in the `Impl TryFrom<BTreeSet<Feature>> for Features {`

**Review by Alex6323 — DISMISSED** [review]

File: sdk/src/types/block/output/feature/mod.rs (hunk 8 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/output/feature/mod.rs:212]: Either way we have to define how we handle already ordered/unique data vs questionable data.

**Review by thibault-martinez — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**